examples/integrations/hf_trainer_api/det_callback.py

```diff
@@ -0,0 +1,348 @@
+import pathlib
+import typing
+
+from transformers import (
+    TrainerCallback,
+    TrainerState,
+    TrainerControl,
+    TrainingArguments,
+)
+
+import os
+import determined as det
+import logging
+import torch
+import uuid
+import weakref
+import importlib.util
+import shutil
+
+
+class DetCallback(TrainerCallback):
+    def __init__(
+        self,
+        args: TrainingArguments,
+        filter_metrics: typing.List[str] = None,
+        tokenizer: typing.Any = None,
+        tokenizer_options: typing.Dict = None,
+        checkpoint_metadata: typing.Dict = None,
+    ) -> None:
+        super().__init__()
+
+        assert (
+            is_determined_available()
+        ), "DetCallback requires determined to be installed. Run `pip install determined`."
+        import determined
+
+        self._det = determined
+        if args.deepspeed:
+            distributed = self._det.core.DistributedContext.from_deepspeed()
+        else:
+            distributed = self._det.core.DistributedContext.from_torch_distributed()
+        self.core_context = self._det.core.init(distributed=distributed)
+        self.core_context.__enter__()
+        weakref.finalize(self, exit_context, self.core_context)
+
+        self.filter_metrics = filter_metrics
+        self.user_checkpoint_metadata = checkpoint_metadata
+        self.tokenizer = tokenizer
+        self.tokenizer_options = tokenizer_options
+        self.load_last_checkpoint(args)
+
+        self.last_metrics: typing.Dict[str, float] = {}
+
+        searcher_config = self._det.get_cluster_info().trial._config["searcher"]
+        self.searcher_metric = searcher_config["metric"]
+        self.searcher_unit = list(searcher_config["max_length"].keys())[0]
+        self.searcher_max_length = list(searcher_config["max_length"].values())[0]
+        self.searcher_ops = self.core_context.searcher.operations()
+        self.current_op = next(self.searcher_ops)
+
+        self._check_searcher_compatibility(args)
+
+    def on_log(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        logs=None,
+        **kwargs,
+    ):
+        if state.is_world_process_zero:
+            metrics, metric_type = self._get_metrics(logs)
+            if metric_type == TRAIN:
+                if (
+                    "train_step" not in self.last_metrics
+                    or self.last_metrics["eval_step"] != state.global_step
```

Comment:
This is done to avoid reporting training/validation metrics twice for the same global_step. AFAIK, this only happens with the **average** training/validation metrics, which, depending on the settings, may be computed for the same global step as regular (not average) metrics leading to DB errors on duplicate key.